### PR TITLE
Add Buffy documentation

### PR DESCRIPTION
**Pull request by halmamol**
This PR includes Buffy documentation in the readthedocs.

It can be checked here: https://pr009-sw-docs.readthedocs.io/en/latest/buffy.html

### CODE REVIEW

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]:
```suggestion
*Sorts MC sensors info into buffers, slays vampires, etc*
```

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]: from the detector maybe better?

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]: effect from the electronics, to be more specific.

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]: into raw waveforms (i.e. with effects from the electronics)

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]: This is not 100% accurate. While `Diomira` uses a realistic model for the PMT electronics, `Hypathia` uses a simplified one (gaussian fluctuations around the baseline +  charge fluctuation for pes).  Both simulate the electronics, but the latter doesn't save the resulting waveforms and feeds them into the pmap creation algorithms.

Perhaps "incomplete electronics modeling" is a better wording, although it is also more obscure...

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]: what does "for each index" mean?

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]: array with shape: (bla, bla, bla).

Also is the length in mus? I thought it was in units of the sampling period.

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]:
```suggestion
     - Maximum duration of the event that will be taken into account starting from the first detected signal. All signals after that are lost. Must be greater than ``buffer_length``. If not, raises a warning and set ``max_time`` == ``buffer_length``
```

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]: is it really an int? It doesn't sound right unless it is in units of the sampling period, but I thought it was in mus. This should also be specified.

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst:74]: The fact that the data from nexus comes binned in time is quite important. It's worth writing it either here as a footnote or somewhere in the introduction. Maybe we should also link to the part of the nexus documentation that specifies how to set this binning (if it exists...)

I've seen now that this is briefly mentioned in the histogram creation section. I think that we should clarify earlier that the binning comes from nexus.

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst:98]: Uff... terrible code organization :rofl:

Comment (gonzaponte) [inline docs/source/buffy.rst:98]:
```suggestion
  Historically, Buffy is based on an initial code of detsim (https://github.com/next-exp/IC/tree/master/invisible_cities/detsim), and most of its functions are located in that path, but they are independent of the :doc:`detsim` city.
```

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]:
```suggestion
Once these histograms are defined, they are sampled according to the binning of each sensor (``pmt_width`` and ``sipm_width``). Sampling widths are included in the simulation parameters (``/MC/info``), and depends on the type of sensor and detector. Normally corresponds to 25 :math:ns for PMTs and and 1 :math:`\mu`\ s for SiPMs.
```

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]: I find this paragraph hard to read, and I think it can be simplified. The relevant bit is that we look for the first value of the binned charge above a certain threshold, which defines our t_trigger. Then, I don't understand what is happening with the PMT/SiPM sum. It's ordered [...] considering [...], which is a touch too generic.

Also, we should mention earlier in the paragraph that this search for charge over threshold happens on the PMT sum.

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]:
```suggestion
Since the ``sensor_width`` is different for each sensor, it is necessary to align and synchronise the clocks between SiPMs and PMTs. Waveforms are sliced then according to binning (``pmt_width`` and ``sipm_width``), trigger time and configured pre-trigger (``pre_trigger``). In addition, it pads with zeros where it is necessary: arrays with zeros where there is no recorder pes in nexus (in the other sensors). Arrays with zeros where no signal and nexus recorded pes otherwise. If more than one trigger is found separated from each other by more than a buffer width, the nexus event can be split into multiple data-like triggers.
```

Comment (gonzaponte) [reply docs/source/buffy.rst]: I think it is important to clarify that we shift the times of the charge histogram such that the first value over threshold falls at t = pre_trigger.

I would rephrase the rest omitting the padding part. It is probably clearer to say that the waveform is an array of zeros filled with the charge from nexus at the corresponding times. The padding is probably an implementation detail that doesn't help the casual reader.

**Review by gonzaponte — COMMENTED** [review]
The images help a lot to visualize the process, good job!

File: docs/source/buffy.rst
Comment (halmamol) [reply docs/source/buffy.rst]: What do you think about something like this?
```suggestion
For detector geometries without sensor electronics, a simplified electronics modeling (gaussian fluctuations around baseline + charge fluctuations for pes) can be applied with :doc:`hypathia` to transform **TWF**\ s into **PMap**\ s.
```

File: docs/source/buffy.rst
Comment (halmamol) [reply docs/source/buffy.rst]: I've been crosschecking, and it's defined as `int` ([here](https://github.com/next-exp/IC/blob/666647c299b559163d47d9b4a17617edfea5886f/invisible_cities/cities/components.py#L991) and here [here](https://github.com/next-exp/IC/blob/666647c299b559163d47d9b4a17617edfea5886f/invisible_cities/detsim/buffer_functions.py#L73)).

File: docs/source/buffy.rst
Comment (gonzaponte) [reply docs/source/buffy.rst]: I now realize that the stuff in parentheses is only part of the story. I am not sure that those two steps are the complete simulation of the PMTs electronics. On top of that, we also ignore the simulation of the SiPMs electronics, which are surely part of `Hypathia`'s job. So perhaps it is better to leave it as "simplified electronics modeling" without the parentheses and refer to the `Hypathia` documentation.

I would also like to keep the word "directly" at the end to indicate that the intermediate waveforms are not stored :)

Comment (halmamol) [reply docs/source/buffy.rst]: Got it! We just need to be sure that it is emphasized on `Hypathia` documentation.

File: docs/source/buffy.rst
Comment (halmamol) [reply docs/source/buffy.rst]: It's not clear, i'll edit it.

File: docs/source/buffy.rst
Comment (halmamol) [reply docs/source/buffy.rst]: You're right! Changed, together with the array.

File: docs/source/buffy.rst
Comment (halmamol) [reply docs/source/buffy.rst:74]: I've added a link to the nexus wiki where the output format is especified and I've extended a little bit the explanation in the Histogram creation section.

File: docs/source/buffy.rst
Comment (halmamol) [reply docs/source/buffy.rst:98]: I've added this to clarify that this detsim folder it is not fully for detsim city. I hope it's clearly explained.

File: docs/source/buffy.rst
Comment (halmamol) [reply docs/source/buffy.rst]: I agree that it could be a little bit confusion, I've removed the info about it and rephrase a little bit earlier in the Histogram creation.

File: docs/source/buffy.rst
Comment (halmamol) [reply docs/source/buffy.rst]: I've rephase it this part a little bit to make it more clear!

File: docs/source/buffy.rst
Comment (gonzaponte) [reply docs/source/buffy.rst:98]: Yes, this was not a comment for you, just a statement for us as IC developers!

**Comment by halmamol** [comment]
I have added the comments, and rephrased it a little bit some parts to make it more clear.

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]:
```suggestion
Full simulations generate sensor (PMTs and SiPMs) responses,
```

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]:
```suggestion
*Buffy* takes nexus sensors' information and sorts it into *true waveforms* (**TWF**).
```

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]: did you mean this?
```suggestion
distortion or effect from the electronics within a certain time interval given by the sensor sampling time.
```

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]:
```suggestion
**TWF**\ s generated by *Buffy* can be transformed into *raw waveforms* **RWF**\ s (i.e. with effects
```

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst:33]: Are these tables generated for the first time in the processing chain here? If so, the documentation should say it explicitly

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst:80]: > and in addition they only provide the information from the sensors and time bins when some charge is detected

Great, I think this sentence is crucial!

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst:87]: Is there a way to put captions on these images? I'm guessing these are examples of the city's output from the context, but it would be great if we could add a caption. Otherwise, we can just put a line above saying that these are some examples.

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]: how about this?
```suggestion
This means that the ``time_bin`` column numbers are increasing for a given event, but they can have gaps since empty time bins are not stored. This initial part of the city checks the time stamp of an event according to the sensors' response and defines histograms of charge distribution
```

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]:
```suggestion
Sampling widths are included in the simulation parameters (``/MC/info``), and depends on the type of sensor and detector. Normally corresponds to 25 :math:`ns` for PMTs and and 1 :math:`\mu s`\  for SiPMs.
```

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]: How about "fills empty bins with zeros" or "restores empty bins by padding zeros in between separate signals"?

Also, we have to be careful with saying "one histogram for pmts and another one for sipms". My feeling is that it is not clear that we perform the syncronization using summed waveforms, but then we generate individual sensor waveforms from that.

Here I would clarify that these are waveforms produced by summing sensors. The generation of individual waveforms should be specified in the section "Synchronisation and trigger separation"

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]:
```suggestion
Once the charge is distributed in the previously defined histograms, the code searches for signal-like events.
```

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]:
```suggestion
It takes the PMT sum histogram and looks for the first value of the binned charge above a certain threshold (``trigger_threshold``), and defines the trigger time, :math:`t_{trigger}`.
```

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]: Here is where I think we need to specify that we generate individual PMT and SiPM waveforms based on the parameters obtained with the summed waveforms.

File: docs/source/buffy.rst
Comment (gonzaponte) [inline docs/source/buffy.rst]: buffer length instead of width to be consistent in the terminology?

Comment (gonzaponte) [reply docs/source/buffy.rst]: Also data-like waveforms instead of triggers?

**Review by gonzaponte — COMMENTED** [review]
Much better. Now it's mostly grammar stuff, but I still think we need to clarify the process of obtaining individual waveforms from the summed waveforms.

File: docs/source/buffy.rst
Comment (halmamol) [reply docs/source/buffy.rst:87]: I already tried to find a way to add captions in read-the-docs and it's not that easy. I'll just add some text below or above.

File: docs/source/buffy.rst
Comment (halmamol) [reply docs/source/buffy.rst:33]: yes, they're! i've included some sentence along the text, not here!

File: docs/source/buffy.rst
Comment (halmamol) [reply docs/source/buffy.rst]: I've included some changes along the text to make this more clear!

**Comment by halmamol** [comment]
Changes included! Before we fix the documentation, we need to sort out the issue with `max_time` since it is defined as an `int` (instead a `float`). I will work on it on IC.

**Comment by gonzaponte** [comment]
I'm happy with this. Let's find someone else to have a second review :)

**Comment by halmamol** [comment]
> I'm happy with this. Let's find someone else to have a second review :)

@gondiaz could give a hand on this, no need to be an expert on the city for the second review!

File: docs/source/buffy.rst
Comment (gondiaz) [inline docs/source/buffy.rst]: set --> sets?

File: docs/source/buffy.rst
Comment (gondiaz) [reply docs/source/buffy.rst]: I dont see any reason for it to be an `int` other than the argument (wrong) typing in the first of the links

File: docs/source/buffy.rst
Comment (gondiaz) [inline docs/source/buffy.rst]: `MC/configuration`

File: docs/source/buffy.rst
Comment (gondiaz) [inline docs/source/buffy.rst]: I think this sentence should be more concrete:
- set the length (in number of samples) as requested in the config parameters (`buffer_length/pmt_width`)

File: docs/source/buffy.rst
Comment (gondiaz) [inline docs/source/buffy.rst]: I would say:
```
to align and synchronise the signals between waveforms
```

Also, remove second "then" in
```
Waveforms are then sliced then according to binning
```

**Review by gondiaz — CHANGES_REQUESTED** [review]

File: docs/source/buffy.rst
Comment (halmamol) [reply docs/source/buffy.rst]: Done

**Review by gondiaz — APPROVED** [review]

**Review by gonzaponte — APPROVED** [review]
The inconsistency in the units of a variable has been addressed. Buffy is documented! Great job, @halmamol!

**Pull request merged**
